src/index/LocatedTriples.cpp

```diff
@@ -0,0 +1,240 @@
+      AD_CORRECTNESS_CHECK(numIndexColumns == 1);
+      return (row[0] > lt->triple_[2]);
+    }
+  };
+  auto cmpEq = [&numIndexColumns](auto lt, auto& row) {
+    if (numIndexColumns == 3) {
+      return (row[0] == lt->triple_[0] && row[1] == lt->triple_[1] &&
+              row[2] == lt->triple_[2]);
+    } else if (numIndexColumns == 2) {
+      return (row[0] == lt->triple_[1] && row[1] == lt->triple_[2]);
+    } else {
+      AD_CORRECTNESS_CHECK(numIndexColumns == 1);
+      return (row[0] == lt->triple_[2]);
+    }
+  };
+  auto writeTripleToResult = [&numIndexColumns, &result](auto resultEntry,
+                                                         auto& locatedTriple) {
+    for (size_t i = 0; i < numIndexColumns; i++) {
+      (*resultEntry)[i] = locatedTriple->triple_[3 - numIndexColumns + i];
+    }
+    // Write UNDEF to any additional columns.
+    for (size_t i = numIndexColumns; i < result.numColumns(); i++) {
+      (*resultEntry)[i] = ValueId::makeUndefined();
+    }
+  };
+
+  for (auto row : block) {
+    // Append triples that are marked for insertion at this `rowIndex` to the
+    // result.
+    while (locatedTriple != locatedTriples.end() && cmpLt(locatedTriple, row)) {
+      if (locatedTriple->shouldTripleExist_ == true) {
+        // Adding an inserted Triple between two triples.
+        writeTripleToResult(resultEntry, locatedTriple);
+        ++resultEntry;
+        ++locatedTriple;
+      } else {
+        // Deletion of a triple that does not exist in the index has no effect.
+        ++locatedTriple;
+      }
+    }
+
+    // Append the triple at this position to the result if and only if it is not
+    // marked for deletion and matches (also skip it if it does not match).
+    bool deleteThisEntry = false;
+    if (locatedTriple != locatedTriples.end() && cmpEq(locatedTriple, row)) {
+      if (locatedTriple->shouldTripleExist_ == false) {
+        // A deletion of a triple that exists in the index.
+        deleteThisEntry = true;
+
+        ++locatedTriple;
+      } else {
+        // An insertion of a triple that already exists in the index has no
+        // effect.
+        ++locatedTriple;
+      }
+    }
+    if (!deleteThisEntry) {
+      *resultEntry++ = row;
+    }
+  }
+  while (locatedTriple != locatedTriples.end() &&
+         locatedTriple->shouldTripleExist_ == true) {
+    writeTripleToResult(resultEntry, locatedTriple);
+    ++resultEntry;
+    ++locatedTriple;
+  }
+
+  // Return the number of rows written to `result`.
+  return resultEntry - (result.begin() + offsetInResult);
+}
+
+// ____________________________________________________________________________
+std::vector<LocatedTriples::iterator> LocatedTriplesPerBlock::add(
+    const std::vector<LocatedTriple>& locatedTriples) {
+  std::vector<LocatedTriples::iterator> handles;
+  handles.reserve(locatedTriples.size());
+  for (auto triple : locatedTriples) {
+    LocatedTriples& locatedTriples = map_[triple.blockIndex_];
```

Comment:
Isn't it more efficient to first sort the triples by the blockIndex and then only do one insert per blockIndex?
Or is that something we will probably update later once we know whether it is a performance bottleneck?

Reply:
Sorting didn't seem that important (little to no improvement) in very early tests with naive sorting. One could sort the triples by one layer up in `DeltaTriples` before the location. Then the locatedTriples should also be sorted automatically. Sorting should probably happens similar to whats done during the index build. But my understanding of these sorters as poor currently. Lets talk about that next week.